### ISSUE TITLE
Bug: Error is thrown in 9.18.0 if unstable_ts_config flag is used

### ISSUE THREAD

**Issue by marcusrbrown** [eslint/eslint#19337]
### Environment

Node version: 22
npm version:
Local ESLint version: 9.18.0
Global ESLint version:
Operating System: Linux

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```
<!-- Paste your configuration here -->
```
</details>

```js
<!-- Paste your code here -->
```
I use Renovate to manage dependencies across several projects. I have been using the unstable_ts_config flag to use eslint.config.ts files for config. While attempting to upgrade to 9.18.0 the Renovate process fails with the error given by ESLint

```sh
Command failed: pnpm run fix

Oops! Something went wrong! :(

ESLint: 9.18.0

Error: The flag 'unstable_ts_config' is inactive: This flag is no longer required to enable TypeScript configuration files.
    at /tmp/renovate/repos/github/bfra-me/renovate-action/node_modules/.pnpm/eslint@9.18.0_jiti@2.4.2/node_modules/eslint/lib/linter/linter.js:1269:23
    at Array.forEach (<anonymous>)
    at new Linter (/tmp/renovate/repos/github/bfra-me/renovate-action/node_modules/.pnpm/eslint@9.18.0_jiti@2.4.2/node_modules/eslint/lib/linter/linter.js:1267:15)
    at new ESLint (/tmp/renovate/repos/github/bfra-me/renovate-action/node_modules/.pnpm/eslint@9.18.0_jiti@2.4.2/node_modules/eslint/lib/eslint/eslint.js:448:24)
    at Object.execute (/tmp/renovate/repos/github/bfra-me/renovate-action/node_modules/.pnpm/eslint@9.18.0_jiti@2.4.2/node_modules/eslint/lib/cli.js:486:24)
    at async main (/tmp/renovate/repos/github/bfra-me/renovate-action/node_modules/.pnpm/eslint@9.18.0_jiti@2.4.2/node_modules/eslint/bin/eslint.js:158:22)
```
You can see the failed job on this PR: https://github.com/bfra-me/renovate-action/pull/1612

### What did you expect to happen?

I expected a deprecation *warning* instead of an error.

### What actually happened?

```sh
Command failed: pnpm run fix

Oops! Something went wrong! :(

ESLint: 9.18.0

Error: The flag 'unstable_ts_config' is inactive: This flag is no longer required to enable TypeScript configuration files.
    at /tmp/renovate/repos/github/bfra-me/renovate-action/node_modules/.pnpm/eslint@9.18.0_jiti@2.4.2/node_modules/eslint/lib/linter/linter.js:1269:23
    at Array.forEach (<anonymous>)
    at new Linter (/tmp/renovate/repos/github/bfra-me/renovate-action/node_modules/.pnpm/eslint@9.18.0_jiti@2.4.2/node_modules/eslint/lib/linter/linter.js:1267:15)
    at new ESLint (/tmp/renovate/repos/github/bfra-me/renovate-action/node_modules/.pnpm/eslint@9.18.0_jiti@2.4.2/node_modules/eslint/lib/eslint/eslint.js:448:24)
    at Object.execute (/tmp/renovate/repos/github/bfra-me/renovate-action/node_modules/.pnpm/eslint@9.18.0_jiti@2.4.2/node_modules/eslint/lib/cli.js:486:24)
    at async main (/tmp/renovate/repos/github/bfra-me/renovate-action/node_modules/.pnpm/eslint@9.18.0_jiti@2.4.2/node_modules/eslint/bin/eslint.js:158:22)
```

### Link to Minimal Reproducible Example

https://github.com/bfra-me/renovate-action/tree/renovate/eslint-monorepo

### Participation

- [x] I am willing to submit a pull request for this issue.

### Additional comments

I understand the flag is marked "unstable" and no longer needed after 9.18.0. However, throwing an error is a breaking change, as I expect to see a deprecation warning and the flag removed in the next major ESLint version.

**Comment by alex-kinokon** [comment]
Broken by #19266.

**Comment by fasttime** [comment]
Thanks for the report @marcusrbrown. This is not a bug, it's intended that using a feature flag like `unstable_ts_config` after it's no longer needed results in an error. Feature flags are used to activate experimental features that could change anytime, also during a minor release like ESLint v9.18.0.

I think in your case you could simply amend the Renovate PR and remove `--flag unstable_ts_config` from the ESLint command:

https://github.com/bfra-me/renovate-action/blob/041f11ba5bec4d080de4ba94497194135d4ee749/package.json#L15 

Am I missing something?

**Comment by marcusrbrown** [comment]
I will amend the dozen repos that cannot be built after 9.18. I disagree that completing the feature behind the flag should mean using the flag results in error - the feature itself remains unchanged between minor versions.

**Comment by nzakas** [comment]
@eslint/eslint-tsc I think it's worth revisiting the behavior of inactive flags. I had previously brought this up on https://github.com/eslint/eslint/issues/19100#issuecomment-2457741994, where I suggested that the migration path of flags from unstable to stable to removed may be a bit too sharp. I think it makes sense to emit a warning in these situations instead of erroring:

* I think it's reasonable that an unstable flag that is removed because the feature has been merged into HEAD.
* Similarly, when a flag moves from unstable to stable (as in `unstable_foo` -> `v10_foo`); when the unstable flag is used it should be treated as if the stable flag is used.
* When a stable flag is removed because its functionality has been merged into HEAD.

If an unstable flag is removed because the functionality is abandoned, then I think it makes sense to throw an error.

**Comment by fasttime** [comment]
> [@eslint/eslint-tsc](https://github.com/orgs/eslint/teams/eslint-tsc) I think it's worth revisiting the behavior of inactive flags. I had previously brought this up on [#19100 (comment)](https://github.com/eslint/eslint/issues/19100#issuecomment-2457741994), where I suggested that the migration path of flags from unstable to stable to removed may be a bit too sharp. I think it makes sense to emit a warning in these situations instead of erroring:
> 
> * I think it's reasonable that an unstable flag that is removed because the feature has been merged into HEAD.
> * Similarly, when a flag moves from unstable to stable (as in `unstable_foo` -> `v10_foo`); when the unstable flag is used it should be treated as if the stable flag is used.
>  * When a stable flag is removed because its functionality has been merged into HEAD.
> 
> If an unstable flag is removed because the functionality is abandoned, then I think it makes sense to throw an error.

Raising warnings for inactive flags was the only expected behavior before #18705. From a user perspective I think that warnings make more sense than errors in most cases because they don't require restricting the version range of the `eslint` dependency when the status of a flag changes, so I'd be in favor of this change. If you want to switch to warnings before the RFC to pass flags as environment variables suggested in #19100, that would be also fine.

**Comment by nzakas** [comment]
> What happens in major releases, are we going to remove inactive flags so that trying to use them at that point will be an error?

I was thinking we could just leave inactive flags in place because they don't take up a lot of memory. I'm not opposed to removing them in major releases if that's what everyone wants.

**Comment by nzakas** [comment]
**TSC Summary:** This issue proposes that we change inactive flags to emit warnings instead of errors in these situations:

* An unstable flag becomes inactive because the feature is now enabled by default.
* A stable flag becomes inactive because the feature is now enabled by default.
* A flag moves from unstable to stable. The unstable flag should work just like the stable flag except that it emits a warning.

If an unstable flag is removed because the functionality is abandoned, then it throws an error.

**TSC Questions:** 

1. Should we accept this proposal?
2. Should inactive flags ever be removed?

**Comment by sam3k** [comment]
Per 2025-01-23 TSC meeting, we've decided:

* We will accept the proposal as described in the issue.
* Inactive flags will be removed each major release.
* Abandoned features that will not land in ESLint should be removed (functionality and flag) in the next minor release.

We will also describe these details on https://eslint.org/docs/latest/flags/ when the new functionality is implemented.

**Comment by jnoordsij** [comment]
Maybe re-opening #19341 might be fruitful then, given that it (attempts to) implement the proposed behavior?

**Comment by nzakas** [comment]
@jnoordsij That PR doesn't represent what was decided on above, so there's no reason to reopen it. If you'd like to start over and implement the new design, please feel free to submit a new PR.

**Comment by jnoordsij** [comment]
@nzakas I don't think I understand what's been decided on then?

>     * We will accept the proposal as described in the issue.
The proposal in the issue, as far as I can tell, is `"I expected a deprecation warning instead of an error."`. 
This was also what you mentioned above:


> **TSC Summary:** This issue proposes that we change inactive flags to emit warnings instead of errors in these situations:
> 
>     * An unstable flag becomes inactive because the feature is now enabled by default.
> 
>     * A stable flag becomes inactive because the feature is now enabled by default.
> 
>     * A flag moves from unstable to stable. The unstable flag should work just like the stable flag except that it emits a warning.

I think that is just what #19341 does? Possibly the only part that might be missing is the implementation for "A flag moves from unstable to stable", given that #19341 does actively filter away all inactive flags. However, I think this is probably something that could be implemented on its own, apart from the work in #19341, particularly because doing such a thing is currently not possible anyways.

If you have any different idea on how an implementation should look, I'd definitely like to hear and maybe even consider working on it. But just to repeat, I currently don't really see how #19341 differs from the proposed solution.

**Comment by nzakas** [comment]
@jnoordsij 

> The proposal in the issue, as far as I can tell, is `"I expected a deprecation warning instead of an error."`.

There are two things your PR is missing:

1. When a flag moves from unstable to stable, we need to both print out a warning and also enable the stable flag.
2. The documentation needs to be updated to describe how flags progress as described in this issue.

We're not going to merge a PR that only does some of what we want, so the options are either to implement everything or we'll let someone else do it.

**Comment by mdjermanovic** [comment]
I'm working on this.
